File: java/videodb-tutorial/src/main/java/com/datastax/videodb/policy/TimeOfDayRetryPolicy.java
package com.datastax.videodb.policy;

import java.util.Calendar;

import com.datastax.driver.core.ConsistencyLevel;
import com.datastax.driver.core.Query;
import com.datastax.driver.core.WriteType;
import com.datastax.driver.core.policies.DowngradingConsistencyRetryPolicy;
import com.datastax.driver.core.policies.RetryPolicy;

public class TimeOfDayRetryPolicy implements RetryPolicy {

	private final DowngradingConsistencyRetryPolicy policy;
	private final int nonRetryStartHour;
	private final int nonRetryEndHour;

	public TimeOfDayRetryPolicy(int nonRetryStartHour, int nonRetryEndHour) {
		policy = DowngradingConsistencyRetryPolicy.INSTANCE;

		this.nonRetryStartHour = nonRetryStartHour;
		this.nonRetryEndHour = nonRetryEndHour;
	}

	@Override
	public RetryDecision onReadTimeout(Query query, ConsistencyLevel cl,
			int requiredResponses, int receivedResponses,
			boolean dataRetrieved, int nbRetry) {

		int hour = Calendar.getInstance().get(Calendar.HOUR);

		if (hour >= nonRetryStartHour && hour <= nonRetryEndHour) {

			return RetryDecision.rethrow();
		}

		return policy.onReadTimeout(query, cl, requiredResponses,
				receivedResponses, dataRetrieved, nbRetry);
	}

	@Override
	public RetryDecision onWriteTimeout(Query query, ConsistencyLevel cl,
			WriteType writeType, int requiredAcks, int receivedAcks, int nbRetry) {
		int hour = Calendar.getInstance().get(Calendar.HOUR);

		if (hour >= nonRetryStartHour && hour <= nonRetryEndHour) {

			return RetryDecision.rethrow();
		}

		return policy.onWriteTimeout(query, cl, writeType, requiredAcks,
				receivedAcks, nbRetry);
	}

	@Override
	public RetryDecision onUnavailable(Query query, ConsistencyLevel cl,
			int requiredReplica, int aliveReplica, int nbRetry) {
		int hour = Calendar.getInstance().get(Calendar.HOUR);

		if (hour >= nonRetryStartHour && hour <= nonRetryEndHour) {

			return RetryDecision.rethrow();
		}

		return policy.onUnavailable(query, cl, requiredReplica, aliveReplica,
				nbRetry);
	}

}

File: csharp/VideoDbApplication/VideoDbApplication/policy/TimeOfDayRetryPolicy.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using Cassandra;

namespace VideoDbApplication.policy
{
    class TimeOfDayRetryPolicy : IRetryPolicy
    {
        private DowngradingConsistencyRetryPolicy policy;
	    private int nonRetryStartHour;
	    private int nonRetryEndHour;

	    public TimeOfDayRetryPolicy(int nonRetryStartHour, int nonRetryEndHour)
        {
		    policy = DowngradingConsistencyRetryPolicy.Instance;
		    this.nonRetryStartHour = nonRetryStartHour;
		    this.nonRetryEndHour = nonRetryEndHour;
	    }

	    public RetryDecision onReadTimeout(Query query, ConsistencyLevel cl,
			    int requiredResponses, int receivedResponses,
			    Boolean dataRetrieved, int nbRetry)
        {
            int hour = DateTime.Now.Hour;
		    if (hour >= nonRetryStartHour && hour <= nonRetryEndHour) {
			    return RetryDecision.Rethrow();
		    }
		    return policy.OnReadTimeout(query, cl, requiredResponses,
				    receivedResponses, dataRetrieved, nbRetry);
	    }

	    public RetryDecision onWriteTimeout(Query query, ConsistencyLevel cl,
			    String writeType, int requiredAcks, int receivedAcks, int nbRetry)
        {
            int hour = DateTime.Now.Hour;
		    if (hour >= nonRetryStartHour && hour <= nonRetryEndHour) {
			    return RetryDecision.Rethrow();
		    }
		    return policy.OnWriteTimeout(query, cl, writeType, requiredAcks,
				    receivedAcks, nbRetry);
	    }

	    public RetryDecision onUnavailable(Query query, ConsistencyLevel cl,
			    int requiredReplica, int aliveReplica, int nbRetry)
        {
            int hour = DateTime.Now.Hour;
		    if (hour >= nonRetryStartHour && hour <= nonRetryEndHour) {
			    return RetryDecision.Rethrow();
		    }
		    return policy.OnUnavailable(query, cl, requiredReplica, aliveReplica,
				    nbRetry);
	    }
    }
}
